builder/azure/arm/step_deploy_template.go

```diff
@@ -93,11 +105,38 @@ func (s *StepDeployTemplate) Cleanup(state multistep.StateBag) {
 	deploymentName := s.name
 	resourceGroupName := state.Get(constants.ArmResourceGroupName).(string)
 	subscriptionId := state.Get(constants.ArmSubscription).(string)
+	deploymentOpsId := deploymentoperations.ResourceGroupDeploymentId{
+		DeploymentName:    deploymentName,
+		ResourceGroupName: resourceGroupName,
+		SubscriptionId:    subscriptionId,
+	}
 	if s.templateType == KeyVaultTemplate {
 		ui.Say("\nDeleting KeyVault created during build")
-		err := s.delete(ctx, subscriptionId, deploymentName, resourceGroupName)
+		deploymentOperations, err := s.listDeploymentOps(ctx, deploymentOpsId)
 		if err != nil {
-			s.reportIfError(err, resourceGroupName)
+			ui.Error(fmt.Sprintf("Could not retrieve deployment operations: %s\n to get the KeyVault, please manually delete it ", err))
+			return
+		}
+		for _, deploymentOperation := range deploymentOperations {
+			// Sometimes an empty operation is added to the list by Azure
+			if deploymentOperation.Properties.TargetResource == nil {
+				continue
+			}
+			resourceName := *deploymentOperation.Properties.TargetResource.ResourceName
+			resourceType := *deploymentOperation.Properties.TargetResource.ResourceType
+
+			if resourceType == KeyVaultResourceType {
+				kvID := commonids.KeyVaultId{
+					VaultName:         resourceName,
+					ResourceGroupName: resourceGroupName,
+					SubscriptionId:    subscriptionId,
+				}
+				err := s.deleteKV(ctx, kvID)
+				if err != nil {
+					s.reportResourceDeletionFailure(err, resourceGroupName)
+				}
+				return
+			}
 		}
 
 	} else {
```

Comment:
This `else` feels like it could be removed if we return at the end of that `if`, since the next action is essentially returning anyway